<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Ideograms | The Core Facility of Bioinformatics’ Blog!</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="Ideograms" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Ideograms in R with ggplot2" />
<meta property="og:description" content="Ideograms in R with ggplot2" />
<link rel="canonical" href="http://localhost:4000/2017/05/28/ideograms.html" />
<meta property="og:url" content="http://localhost:4000/2017/05/28/ideograms.html" />
<meta property="og:site_name" content="The Core Facility of Bioinformatics’ Blog!" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2017-05-28T00:00:00+02:00" />
<script type="application/ld+json">
{"description":"Ideograms in R with ggplot2","@type":"BlogPosting","headline":"Ideograms","dateModified":"2017-05-28T00:00:00+02:00","datePublished":"2017-05-28T00:00:00+02:00","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/2017/05/28/ideograms.html"},"url":"http://localhost:4000/2017/05/28/ideograms.html","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="The Core Facility of Bioinformatics' Blog!" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">The Core Facility of Bioinformatics&#39; Blog!</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/about/">About</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Ideograms</h1>
    <p class="post-meta">
      <time class="dt-published" datetime="2017-05-28T00:00:00+02:00" itemprop="datePublished">May 28, 2017
      </time></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h1 id="ideograms-in-r-with-ggplot2">Ideograms in R with ggplot2</h1>

<h3 id="introduction">Introduction</h3>

<p>Ideograms show the chromosomes of a selected genome with a nice banding pattern that has been yielded from different staining techniques in “real” wetlab biological experiments.
The banding pattern shows regions of condendensed and open DNA.
In bioinformatics this makes for a good representation of chromosomes compared to dull x-axis counting from 0Mb to end of chromosome.</p>

<p><img src="figures/2017-05-28-ideograms_fig1.png" alt="dull plot along the chromosome" /></p>
<div class="captionheader">Figure 1:<div class="caption">p-values plotted along chromosomes as plain x-axis coordinates. A rather sad view, though the values are plotted well.</div></div>

<p>An ideogram would be a nice addition to make the x-axis more fancy and in addition stack the chromosomes more tightly instead of using <code class="highlighter-rouge">facet_wrap</code> from the ggplot2 package.</p>

<p><img src="figures/2017-05-28-ideograms_fig2.png" alt="ideograms of the human genome" /></p>
<div class="captionheader">Figure 2:<div class="caption">Chromosomes depicted as ideograms.</div></div>

<p>To plot data, e.g. coverage tracks or calculated p-values, along the genome one needs to add the respective track of information to the ideogram.</p>

<h3 id="solution---first-steps">Solution - first steps</h3>

<p>For a beginning one could follow the Biostars question of Alexander Skates (https://www.biostars.org/p/152969/).
This will guide you through the setup of the respective data frames and <code class="highlighter-rouge">GenomicRanges</code> object.</p>

<p>First setup your R session to know about the right packages.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"ggbio"</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="s2">"GenomicRanges"</span><span class="p">)</span><span class="w">
</span><span class="n">require</span><span class="p">(</span><span class="s2">"biovizBase"</span><span class="p">)</span><span class="w">

</span><span class="c1"># some test data</span><span class="w">
</span><span class="n">CHROM</span><span class="w">    </span><span class="n">POS</span><span class="w">  </span><span class="n">fisher_het</span><span class="w">
</span><span class="m">1</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">134775</span><span class="w"> </span><span class="m">0.299587633</span><span class="w">
</span><span class="m">2</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">135237</span><span class="w"> </span><span class="m">1.000000000</span><span class="w">
</span><span class="m">3</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">135277</span><span class="w"> </span><span class="m">0.483198279</span><span class="w">
</span><span class="m">4</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">135331</span><span class="w"> </span><span class="m">0.224587437</span><span class="w">
</span><span class="m">5</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">135334</span><span class="w"> </span><span class="m">0.068035761</span><span class="w">
</span><span class="m">6</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">135656</span><span class="w"> </span><span class="m">0.468998144</span><span class="w">
</span><span class="m">7</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">135708</span><span class="w"> </span><span class="m">0.746611845</span><span class="w">
</span><span class="m">8</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">135801</span><span class="w"> </span><span class="m">0.242257762</span><span class="w">
</span><span class="m">9</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">135853</span><span class="w"> </span><span class="m">0.001234701</span><span class="w">
</span><span class="m">10</span><span class="o">:</span><span class="w">    </span><span class="m">10</span><span class="w"> </span><span class="m">137186</span><span class="w"> </span><span class="m">0.774670848</span><span class="w">
</span></code></pre></div></div>

<p>Now we need to setup the object containing the banding pattern and a <code class="highlighter-rouge">GRanges</code> object to contain your data.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load banding data</span><span class="w">
</span><span class="n">data</span><span class="p">(</span><span class="n">hg19IdeogramCyto</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"biovizBase"</span><span class="p">)</span><span class="w">
</span><span class="n">hg19</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keepSeqlevels</span><span class="p">(</span><span class="n">hg19IdeogramCyto</span><span class="p">,</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"chr"</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">,</span><span class="w"> </span><span class="s2">"X"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Y"</span><span class="p">)))</span><span class="w">

</span><span class="c1"># create a test GRanges object</span><span class="w">
</span><span class="c1"># from the test data given above</span><span class="w">
</span><span class="n">test.granges</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GRanges</span><span class="p">(</span><span class="n">seqnames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"chr"</span><span class="p">,</span><span class="w"> </span><span class="n">df.test.data</span><span class="o">$</span><span class="n">CHROM</span><span class="p">),</span><span class="w">
                        </span><span class="n">ranges</span><span class="o">=</span><span class="n">IRanges</span><span class="p">(</span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">df.test.data</span><span class="o">$</span><span class="n">POS</span><span class="p">,</span><span class="w">
                                       </span><span class="n">end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">df.test.data</span><span class="o">$</span><span class="n">POS</span><span class="p">),</span><span class="w">
                        </span><span class="n">strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"*"</span><span class="p">,</span><span class="w">
                        </span><span class="n">fisher_het</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">df.test.data</span><span class="o">$</span><span class="n">fisher_het</span><span class="p">)</span><span class="w">

</span><span class="c1"># attach chromosome lengths</span><span class="w">
</span><span class="n">data</span><span class="p">(</span><span class="n">hg19Ideogram</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"biovizBase"</span><span class="p">)</span><span class="w">
</span><span class="n">seqlengths</span><span class="p">(</span><span class="n">test.granges</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">seqlengths</span><span class="p">(</span><span class="n">hg19Ideogram</span><span class="p">)[</span><span class="nf">names</span><span class="p">(</span><span class="n">seqlengths</span><span class="p">(</span><span class="n">test.granges</span><span class="p">))]</span><span class="w">

</span></code></pre></div></div>

<h3 id="solution---the-final-steps">Solution - the final steps</h3>

<p>The solution of Alexander Skates is only short by a few meters, if one wants to add a second track of data to the ideograms and doesn’t really know how to handle <code class="highlighter-rouge">GenomicRanges</code> objects like me.
Luckily there’s help underway found here http://www.tengfei.name/ggbio/docs/man/layout_karyogram-method.html.
The critical part here is that it points out how to use the <code class="highlighter-rouge">GRanges</code> object correctly and use the coordinates stored in the object using <code class="highlighter-rouge">start</code>.
And there we go.</p>

<ul>
  <li>the x-coordinate from the <code class="highlighter-rouge">GRanges</code> object to be used would be <code class="highlighter-rouge">start</code></li>
  <li><code class="highlighter-rouge">ylim</code> defines the size of the point-subplots
    <ul>
      <li>play around with those scales, if you - for example - want to shift the lower points behind the ideograms</li>
    </ul>
  </li>
  <li><code class="highlighter-rouge">size</code> needs to be reduced a little as default point size would produce a nearly unreadable plot</li>
  <li><code class="highlighter-rouge">geom</code> can contain anything that <code class="highlighter-rouge">ggplot2</code> would understand as <code class="highlighter-rouge">geom_...</code></li>
</ul>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">hg19</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
     </span><span class="n">layout_karyogram</span><span class="p">(</span><span class="n">cytoband</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
     </span><span class="n">layout_karyogram</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test.granges</span><span class="p">,</span><span class="w">
                      </span><span class="n">geom</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"point"</span><span class="p">,</span><span class="w">
                      </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">start</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">fisher_het</span><span class="p">),</span><span class="w">
                      </span><span class="n">ylim</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">10</span><span class="p">,</span><span class="m">50</span><span class="p">),</span><span class="w">
                      </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">,</span><span class="w">
                      </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.4</span><span class="w">
                      </span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="figures/2017-05-28-ideograms_fig3.png" alt="ideograms of the human genome with p-values attached" /></p>
<div class="captionheader">Figure 3:<div class="caption">p-values plotted along chromosomes depicted as ideograms. The exact values shown as points cannot be well read though.</div></div>

<h3 id="discussion">Discussion</h3>

<p>The process of finding this solution can be found in a <a href="https://stackoverflow.com/questions/44003072/annotate-karyogram-with-granges-track/44043471#44043471">Stackoverflow post</a>. <br />
This plot gives a good overview of your data along the chromosomes to show where there might be significant peaks and bumps in your data.
The scale of the y-axis is a bit lost and if one is interested in this one may need to resort to the dull plot in figure 1 or find a better way to add tracks to ideograms, which I would gladly add here.</p>

<h3 id="supplements">Supplements</h3>

<p>PS: The R code to produce figure 1 and 2</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># figure 1</span><span class="w">

</span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">my.data.frame</span><span class="p">,],</span><span class="w">
       </span><span class="n">aes</span><span class="p">(</span><span class="n">POS</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">fisher_het</span><span class="p">))</span><span class="w">
       </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">CHROM</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Position on chromosome"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"-log(raw p-Value)"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">geom_hline</span><span class="p">(</span><span class="n">yintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w">

</span><span class="c1"># figure 2</span><span class="w">

</span><span class="n">ggplot</span><span class="p">(</span><span class="n">hg19</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">layout_karyogram</span><span class="p">(</span><span class="n">cytoband</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>

  </div><a class="u-url" href="/2017/05/28/ideograms.html" hidden></a>
</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">The Core Facility of Bioinformatics&#39; Blog!</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">The Core Facility of Bioinformatics&#39; Blog!</li><li><a class="u-email" href="mailto:imb-cf-bioinformatics (at) imb-mainz.de">imb-cf-bioinformatics (at) imb-mainz.de</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/imbforge"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">imbforge</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>This blog contains our thoughts and discoveries, and serves as a whiteboard to the outside world.
</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
