<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Merge Track Files | The Core Facility of Bioinformatics’ Blog!</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="Merge Track Files" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Merge track files" />
<meta property="og:description" content="Merge track files" />
<link rel="canonical" href="http://localhost:4000/2016/11/04/Merge-track-files.html" />
<meta property="og:url" content="http://localhost:4000/2016/11/04/Merge-track-files.html" />
<meta property="og:site_name" content="The Core Facility of Bioinformatics’ Blog!" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2016-11-04T00:00:00+01:00" />
<script type="application/ld+json">
{"description":"Merge track files","@type":"BlogPosting","headline":"Merge Track Files","dateModified":"2016-11-04T00:00:00+01:00","datePublished":"2016-11-04T00:00:00+01:00","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/2016/11/04/Merge-track-files.html"},"url":"http://localhost:4000/2016/11/04/Merge-track-files.html","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="The Core Facility of Bioinformatics' Blog!" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">The Core Facility of Bioinformatics&#39; Blog!</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/about/">About</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Merge Track Files</h1>
    <p class="post-meta">
      <time class="dt-published" datetime="2016-11-04T00:00:00+01:00" itemprop="datePublished">Nov 4, 2016
      </time></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h1 id="merge-track-files">Merge track files</h1>

<h2 id="bigwig-tracks">bigWig tracks</h2>

<p>Generating a single track summarizing NGS coverage from several replicates is a common request among scientists.</p>

<p>Usually, the scientist will be happy if the bioinformatician just merges together the <em>coverage per million</em> <strong>normalized</strong> tracks. This is a trivial task handled with the tools from the guys of the UCSC Genome Bioinformatics Group (aka <a href="https://genome.ucsc.edu/util.html"><em>the kent utilities</em></a>).</p>

<p>All the magic is done with the <code class="highlighter-rouge">bigWigMerge</code> tool, which puts together the signal from the several bigwig tracks, (bedGraph output), to be eventually converted back to bigwig with the <code class="highlighter-rouge">bedGraphToBigWig</code> tool. The <em>trick</em> here is to divide to amount of signal in each position by the number of replicates. Something like this will do the job:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bigWigMerge sample1_rep1.bw sample1_rep2.bw sample1_rep3.bw stdout <span class="se">\</span>
    | awk <span class="nt">-v</span> <span class="nv">NORM</span><span class="o">=</span>1/3 <span class="s1">'$4=(NORM)*$4'</span> <span class="se">\</span>
    | <span class="nv">LC_COLLATE</span><span class="o">=</span>C sort <span class="nt">-k1</span>,1 <span class="nt">-k2</span>,2n <span class="o">&gt;</span> sample1.bed
bedGraphToBigWig sample1.bed chr.sizes sample1.bw <span class="o">&amp;&amp;</span> rm sample1.bed
</code></pre></div></div>

<p>Chromosome sizes can be retrieved from the UCSC Genome Browser tools with the tool <code class="highlighter-rouge">fetchChromSizes</code>, also from the kent utilities. Alternatively, <code class="highlighter-rouge">samtools idxstats sample1.bam | cut -f1-2 &gt; ./chr.sizes</code> will work.</p>

<h3 id="requirements">Requirements</h3>

<ul>
  <li><code class="highlighter-rouge">bedGraphToBigWig</code> from the <a href="https://genome.ucsc.edu/util.html">kent utilities</a>.</li>
  <li>the cromosome sizes, which can be retrieved from the UCSC Genome Browser tools with the fetchChromSizes. Alternatively, <code class="highlighter-rouge">samtools idxstats sample1.bam | cut -f1-2 &gt; ./chr.sizes</code>.</li>
</ul>

<h3 id="source">Source</h3>

<p>The following script loops over serveral samples defined in a array, and identifies the replicates assuming they follow a pattern (matched by the ?? wildcards):</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>
<span class="nv">UCSC</span><span class="o">=</span>/opt/ucsc/latest/
<span class="nv">INPUTDIR</span><span class="o">=</span>/project/tracks
<span class="nv">SAMPLES</span><span class="o">=(</span>sample1_?? sample2_?? sample3_??<span class="o">)</span>
<span class="nv">CHRSIZES</span><span class="o">=</span>/project/annotation/chr.sizes  <span class="c"># got with: samtools idxstats sample1.bam | cut -f1-2 &gt; ./chr.sizes</span>

<span class="nb">cd</span> <span class="nv">$INPUTDIR</span>

<span class="k">for </span>SAMPLE <span class="k">in</span> <span class="k">${</span><span class="nv">SAMPLES</span><span class="p">[@]</span><span class="k">}</span><span class="p">;</span> <span class="k">do
    </span><span class="nb">echo</span> <span class="nv">$SAMPLE</span>

    <span class="nv">REP</span><span class="o">=</span><span class="k">$(</span><span class="nb">ls</span> <span class="k">${</span><span class="nv">SAMPLE</span><span class="k">}*</span>.bw<span class="k">)</span> <span class="c"># get list of replicates to merge</span>
    <span class="nv">N</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span> <span class="nv">$REP</span> | wc <span class="nt">-w</span><span class="k">)</span>  <span class="c"># how many?</span>
    <span class="nv">NORM</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span> <span class="s2">"1 / </span><span class="nv">$N</span><span class="s2">"</span> | bc <span class="nt">-l</span><span class="k">)</span>   <span class="c"># average each signal by the number of replicates</span>
    <span class="nv">OUT</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span> <span class="nv">$SAMPLE</span> | sed <span class="s1">'s/??_//'</span><span class="k">)</span>

    <span class="k">${</span><span class="nv">UCSC</span><span class="k">}</span>/bigWigMerge <span class="nv">$REP</span> stdout <span class="se">\ </span>                   <span class="c"># merge replicates and output to stdout</span>
        | awk <span class="nt">-v</span> <span class="nv">NORM</span><span class="o">=</span><span class="nv">$NORM</span> <span class="s1">'$4=NORM*$4'</span> <span class="se">\ </span>              <span class="c"># normalize (4th field contains the coverage)</span>
        | <span class="nv">LC_COLLATE</span><span class="o">=</span>C sort <span class="nt">-k1</span>,1 <span class="nt">-k2</span>,2n <span class="o">&gt;</span> <span class="k">${</span><span class="nv">OUT</span><span class="k">}</span>.bed    <span class="c"># sort by chr-pos</span>
    <span class="k">${</span><span class="nv">UCSC</span><span class="k">}</span>/bedGraphToBigWig <span class="k">${</span><span class="nv">OUT</span><span class="k">}</span>.bed <span class="nv">$CHRSIZES</span> <span class="k">${</span><span class="nv">OUT</span><span class="k">}</span>.bw <span class="o">&amp;&amp;</span> rm <span class="k">${</span><span class="nv">OUT</span><span class="k">}</span>.bed
<span class="k">done</span>
</code></pre></div></div>

<h2 id="bam-files">BAM files</h2>

<p>Sometimes the researcher might be interested in keeping information at read level, e.g. to explore splicing. In this case, usual formats to represent continuous data in tracks are of no use, and the bioinformatician has to fall back to BAM.</p>

<p>The main basic thing to consider is avoid overrepresentation of any library over the rest. Thus, we have to first calculate the library size of the replicates, and subsample from the original bam files to the smallest fraction before merging.</p>

<h3 id="steps">Steps</h3>

<ul>
  <li>get the smallest library size of all replicates of the sample</li>
  <li>downsample the rest of the replicates to the smallest library</li>
  <li>merge the downsampled replicates together</li>
</ul>

<h3 id="requirements-1">Requirements</h3>

<ul>
  <li><a href="http://www.htslib.org/">samtools</a></li>
</ul>

<h3 id="source-1">Source</h3>

<p>The following script loops over serveral samples defined in a array, and identifies the replicates assuming they follow a pattern (matched by the ?? wildcards):</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c">#!/bin/bash</span>
<span class="nv">SAMTOOLS</span><span class="o">=</span>/opt/samtools/1.3/samtools
<span class="nv">INPUTDIR</span><span class="o">=</span>/project/mapped
<span class="nv">SAMPLES</span><span class="o">=(</span>sample1_?? sample2_?? sample3??<span class="o">)</span>
<span class="nv">SEED</span><span class="o">=</span>666

<span class="nb">cd</span> <span class="nv">$INPUTDIR</span>

<span class="k">for </span>SAMPLE <span class="k">in</span> <span class="k">${</span><span class="nv">SAMPLES</span><span class="p">[@]</span><span class="k">}</span><span class="p">;</span> <span class="k">do
    </span><span class="nv">OUT</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span> <span class="nv">$SAMPLE</span> | sed <span class="s1">'s/??_//'</span><span class="k">)</span>
    <span class="nb">echo</span> <span class="s2">"== </span><span class="nv">$OUT</span><span class="s2"> =="</span>

    <span class="c"># get minimum number of reads mapped</span>
    <span class="nv">REPS</span><span class="o">=()</span> <span class="c"># will fill out later</span>
    <span class="nv">MAPPEDREADS</span><span class="o">=()</span>  <span class="c"># will fill out later</span>
    <span class="k">for </span>REP <span class="k">in</span> <span class="k">${</span><span class="nv">SAMPLE</span><span class="k">}*</span>.bam<span class="p">;</span> <span class="k">do
        </span><span class="nv">M</span><span class="o">=</span><span class="k">$(</span><span class="nv">$SAMTOOLS</span> flagstat <span class="nv">$REP</span> | <span class="nb">grep </span>mapped | cut <span class="nt">-f1</span> <span class="nt">-d</span><span class="s2">" "</span> | head <span class="nt">-n1</span><span class="k">)</span>
        MAPPEDREADS+<span class="o">=(</span><span class="nv">$M</span><span class="o">)</span>
        REPS+<span class="o">=(</span><span class="nv">$REP</span><span class="o">)</span>
        <span class="nb">echo</span> <span class="s2">"</span><span class="nv">$REP</span><span class="s2"> has </span><span class="nv">$M</span><span class="s2"> reads"</span>
    <span class="k">done
    </span><span class="nv">MIN</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span> <span class="k">${</span><span class="nv">MAPPEDREADS</span><span class="p">[@]</span><span class="k">}</span> | tr <span class="s2">" "</span> <span class="s2">"</span><span class="se">\n</span><span class="s2">"</span> | sort <span class="nt">-n</span> | head <span class="nt">-n1</span><span class="k">)</span>

    <span class="c"># downsample to the minimum</span>
    <span class="nv">REPSDOWNSAMPLED</span><span class="o">=()</span>
    <span class="k">for </span>i <span class="k">in</span> <span class="k">$(</span>seq 1 <span class="k">${#</span><span class="nv">REPS</span><span class="p">[@]</span><span class="k">})</span><span class="p">;</span> <span class="k">do
        </span><span class="nv">X</span><span class="o">=</span><span class="k">${</span><span class="nv">REPS</span><span class="p">[</span><span class="nv">$i</span><span class="p">-1]</span><span class="k">}</span>
        REPSDOWNSAMPLED+<span class="o">=(</span><span class="k">${</span><span class="nv">X</span><span class="p">%.bam</span><span class="k">}</span>.subsampled.bam<span class="o">)</span>
        <span class="k">if</span> <span class="o">[</span> <span class="s2">"</span><span class="nv">$MIN</span><span class="s2">"</span> <span class="nt">-eq</span> <span class="s2">"</span><span class="k">${</span><span class="nv">MAPPEDREADS</span><span class="p">[</span><span class="nv">$i</span><span class="p">-1]</span><span class="k">}</span><span class="s2">"</span> <span class="o">]</span><span class="p">;</span> <span class="k">then
            </span><span class="nb">echo</span> <span class="s2">"not downsampling </span><span class="nv">$X</span><span class="s2">"</span>
            <span class="nv">$SAMTOOLS</span> view <span class="nt">-F4</span> <span class="nt">-bh</span> <span class="nv">$X</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">REPSDOWNSAMPLED</span><span class="p">[</span><span class="nv">$i</span><span class="p">-1]</span><span class="k">}</span>
        <span class="k">else
            </span><span class="nv">PROB</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span> <span class="s2">"</span><span class="nv">$MIN</span><span class="s2"> / </span><span class="k">${</span><span class="nv">MAPPEDREADS</span><span class="p">[</span><span class="nv">$i</span><span class="p">-1]</span><span class="k">}</span><span class="s2">"</span> | bc <span class="nt">-l</span> | cut <span class="nt">-c2-3</span><span class="k">)</span>
            <span class="nb">echo</span> <span class="s2">"downsampling </span><span class="nv">$X</span><span class="s2"> to </span><span class="k">${</span><span class="nv">PROB</span><span class="k">}</span><span class="s2">% of reads"</span>
            <span class="nv">$SAMTOOLS</span> view <span class="nt">-s</span> <span class="k">${</span><span class="nv">SEED</span><span class="k">}</span>.<span class="k">${</span><span class="nv">PROB</span><span class="k">}</span> <span class="nt">-F4</span> <span class="nt">-bh</span> <span class="nv">$X</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">REPSDOWNSAMPLED</span><span class="p">[</span><span class="nv">$i</span><span class="p">-1]</span><span class="k">}</span>
        <span class="k">fi
    done</span>

    <span class="c"># merge the downsampled bams</span>
    <span class="nb">echo</span> <span class="s2">"merging </span><span class="k">${</span><span class="nv">REPSDOWNSAMPLED</span><span class="p">[@]</span><span class="k">}</span><span class="s2"> into </span><span class="k">${</span><span class="nv">OUT</span><span class="k">}</span><span class="s2">.bam"</span>
    <span class="nv">$SAMTOOLS</span> merge <span class="k">${</span><span class="nv">OUT</span><span class="k">}</span>.bam <span class="k">${</span><span class="nv">REPSDOWNSAMPLED</span><span class="p">[@]</span><span class="k">}</span> <span class="o">&amp;&amp;</span> rm <span class="k">${</span><span class="nv">REPSDOWNSAMPLED</span><span class="p">[@]</span><span class="k">}</span>
<span class="k">done</span>
</code></pre></div></div>

<p>And happy sashimi ;-)</p>

<p><img src="figures/2016-11-04-Merge-track-files_fig1.png" alt="happy sashimi!" /></p>

  </div><a class="u-url" href="/2016/11/04/Merge-track-files.html" hidden></a>
</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">The Core Facility of Bioinformatics&#39; Blog!</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">The Core Facility of Bioinformatics&#39; Blog!</li><li><a class="u-email" href="mailto:imb-cf-bioinformatics (at) imb-mainz.de">imb-cf-bioinformatics (at) imb-mainz.de</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/imbforge"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">imbforge</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>This blog contains our thoughts and discoveries, and serves as a whiteboard to the outside world.
</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
